File: backend/openai_client.py
from typing import List, Dict, Any, Optional
import openai
from config import Config

class Message:
    def __init__(self, role: str, content: str):
        self.role = role
        self.content = content

    def to_dict(self) -> Dict[str, str]:
        return {
            "role": self.role,
            "content": self.content
        }

class LLMClient:
    def __init__(self):
        # # Clear any proxy settings that may be causing issues
        # import os
        # if 'http_proxy' in os.environ:
        #     del os.environ['http_proxy']
        # if 'https_proxy' in os.environ:
        #     del os.environ['https_proxy']
        # if 'HTTP_PROXY' in os.environ:
        #     del os.environ['HTTP_PROXY']
        # if 'HTTPS_PROXY' in os.environ:
        #     del os.environ['HTTPS_PROXY']
        
        # self.openai_client = openai.OpenAI(api_key=Config.OPENAI_API_KEY)
        import os
        for proxy_var in ['http_proxy', 'https_proxy', 'HTTP_PROXY', 'HTTPS_PROXY']:
            if proxy_var in os.environ:
                del os.environ[proxy_var]
        
        # Create client with minimal configuration
        import openai
        from openai import OpenAI
        
        try:
            # Try creating with just the API key
            self.openai_client = openai.OpenAI(api_key=Config.OPENAI_API_KEY)
        except TypeError as e:
            # If that fails, print diagnostic info and try alternative approach
            print(f"Error initializing OpenAI client: {str(e)}")
            print(f"OpenAI version: {openai.__version__}")
            
            # Alternative initialization with no keyword args
            from openai import OpenAI
            self.openai_client = OpenAI()
            # Set API key after initialization
            self.openai_client.api_key = Config.OPENAI_API_KEY
    
    async def generate_response(self, messages: List[Dict[str, str]], model: str) -> Dict[str, Any]:
        """
        Generate a response using the specified model.
        """
        try:
            return await self._call_openai_api(messages, model)
        except Exception as e:
            return {
                "error": str(e),
                "message": {
                    "role": "assistant",
                    "content": f"Sorry, there was an error: {str(e)}"
                }
            }
    
    async def _call_openai_api(self, messages: List[Dict[str, str]], model: str) -> Dict[str, Any]:
        """
        Call the OpenAI API to generate a response.
        """
        try:
            print(f"Calling OpenAI API with model: {model}")
            print(f"Using API Key: {Config.OPENAI_API_KEY[:5]}..." if Config.OPENAI_API_KEY else "No API key found!")
            
            # For newer OpenAI library (v1.0.0+)
            response = self.openai_client.chat.completions.create(
                model=model,
                messages=messages,
                max_completion_tokens=Config.MAX_COMPLETION_TOKENS,
                # temperature=Config.TEMPERATURE
            )
            
            # No need to await here in newer versions
            return {
                "message": {
                    "role": "assistant",
                    "content": response.choices[0].message.content
                }
            }
        except Exception as e:
            print(f"ERROR in OpenAI API call: {str(e)}")
            import traceback
            traceback.print_exc()
            raise

File: chat-interface/lib/openai-client.ts
// OpenAI client for the application
// Replaces backend/openai_client.py

import OpenAI from 'openai';
import { Config } from './config';
import { Message } from './types';

export class OpenAIClient {
  private client: OpenAI;
  
  constructor() {
    // Create the OpenAI client with API key
    this.client = new OpenAI({
      apiKey: Config.OPENAI_API_KEY,
    });
  }
  
  /**
   * Generate a response using the specified model
   * @param messages The chat history
   * @param model The model to use
   * @returns The generated response
   */
  async generateResponse(messages: Message[], model: string): Promise<{
    message: Message;
    error?: string;
  }> {
    try {
      return await this.callOpenAIAPI(messages, model);
    } catch (error) {
      console.error('Error in generateResponse:', error);
      
      return {
        message: {
          role: 'assistant',
          content: `Sorry, there was an error: ${error instanceof Error ? error.message : 'Unknown error occurred'}`
        },
        error: error instanceof Error ? error.message : 'Unknown error occurred'
      };
    }
  }
  
  /**
   * Call the OpenAI API to generate a response
   * @param messages The chat history
   * @param model The model to use
   * @returns The generated response
   */
  private async callOpenAIAPI(messages: Message[], model: string): Promise<{
    message: Message;
    error?: string;
  }> {
    try {
      console.log(`Calling OpenAI API with model: ${model}`);
      
      // Convert our Message format to OpenAI's ChatCompletionMessageParam
      const openAIMessages = messages.map(msg => ({
        role: msg.role as 'user' | 'assistant' | 'system',
        content: msg.content
      }));
      
      const response = await this.client.chat.completions.create({
        model: model,
        messages: openAIMessages,
        max_tokens: Config.MAX_COMPLETION_TOKENS,
      });
      
      // Return the response content
      return {
        message: {
          role: 'assistant',
          content: response.choices[0].message.content || ''
        }
      };
    } catch (error) {
      console.error('ERROR in OpenAI API call:', error);
      throw error;
    }
  }
}